### PR TITLE
applications: nrf_desktop: Link doxygen API to rst

### PR DESCRIPTION
**Pull request by pdunaj**
Ensure that API documented with doxygen is visible in rst files.
Add missing documentation.

Jira: NCSDK-14306

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Comment by pdunaj** [comment]
Hi @divipillai , please check if this is what you are looking for.

**Review by divipillai — COMMENTED** [review]
Hi Pawel,
Yes, this is what is expected. I have added a few review comments.

File: applications/nrf_desktop/README.rst [hunk 1]
```diff
@@ -1831,6 +1831,43 @@ These are valid for events that have many listeners or sources, and are gathered
    doc/hfclk_lock.rst
    doc/event_rel_modules.rst
 
+Application configuration API
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1834]:
```suggestion
API documentation
```

File: applications/nrf_desktop/README.rst [hunk 2]
```diff
@@ -1831,6 +1831,43 @@ These are valid for events that have many listeners or sources, and are gathered
...
+*****************************
+
+Listed below are elements of API used by application configuration.
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1837]:
```suggestion
Following are the API elements used by the application configuration.
```

File: applications/nrf_desktop/configuration/common/hid_report_desc.h [hunk 3]
```diff
@@ -23,28 +27,36 @@
...
+/**
+ * @brief HID Reports
+ * @defgroup nrf_desktop_hid_reports HID Reports
+ * @{
+ */
+
+/** @brief Identification numbers of HID reports. */
 enum report_id {
-	REPORT_ID_RESERVED,
+	REPORT_ID_RESERVED,		/**< Reserved. */
```
Comment (divipillai) [inline applications/nrf_desktop/configuration/common/hid_report_desc.h:38]: The comment /**< Reserved. */ should be before REPORT_ID_RESERVED.
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#enums

Same applies for all the parameters.

File: applications/nrf_desktop/README.rst [hunk 4]
```diff
@@ -1831,6 +1831,43 @@ These are valid for events that have many listeners or sources, and are gathered
...
+
+HID reports
+===========
+
+| Header file: :file:`applications/nrf_desktop/configuration/common/hid_report_desc.h`
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1842]: Please give the details of c files also: For example, Source files: file:`applications/nrf_desktop/configuration/common/hid_report_desc.c`

Also, in all the below cases.

File: applications/nrf_desktop/README.rst [hunk 5]
```diff
@@ -1831,6 +1831,43 @@ These are valid for events that have many listeners or sources, and are gathered
...
+| Header file: :file:`applications/nrf_desktop/configuration/common/led_state.h`
+
+.. doxygengroup:: nrf_desktop_led_state
+   :project: nrf
+   :members:
+
+Application events API
+**********************
+
+Listed below are events grouped by module.
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1860]:
```suggestion
Following are the events grouped by module.
```

File: applications/nrf_desktop/README.rst (hunk 4 above)
Comment (pdunaj) [reply applications/nrf_desktop/README.rst:1842]: Do we really want to do it? I personally would try to expose only the API here not the details of implementation.

**Comment by pdunaj** [comment]
rebased

File: applications/nrf_desktop/README.rst (hunk 4 above)
Comment (divipillai) [reply applications/nrf_desktop/README.rst:1851]: I mean to provide the Source file name. For e.g: http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-7484/nrf/applications/asset_tracker_v2/doc/util_module.html#api-documentation

image

**Comment by pdunaj** [comment]
Actual changes

**Comment by pdunaj** [comment]
pushing fix for bad sign-off. It seems checkpatch started to reject the `co-authored-by` tag

File: applications/nrf_desktop/README.rst (hunk 4 above)
Comment (pdunaj) [reply applications/nrf_desktop/README.rst:1851]: should be ok now

File: applications/nrf_desktop/src/events/usb_event.h [hunk 6]
```diff
@@ -26,40 +20,59 @@
...
+ * @{
+ */
 
 /** @brief Peer states. */
 enum usb_state {
-	USB_STATE_DISCONNECTED, /**< Cable is not attached. */
-	USB_STATE_POWERED,      /**< Cable attached but no communication. */
-	USB_STATE_ACTIVE,       /**< Cable attached and data is exchanged. */
-	USB_STATE_SUSPENDED,    /**< Cable attached but port is suspended. */
+	/** Cable is not attached. */
```
Comment (rakons) [inline applications/nrf_desktop/src/events/usb_event.h:35]: Just an opinion - the inline version is easier to read.

**Review by rakons — APPROVED** [review]
Generally I found inline enumerator documentation easier to read as i allows me to have more lines on the screen. Not a blocker.

File: applications/nrf_desktop/src/events/usb_event.h (hunk 6 above)
Comment (pdunaj) [reply applications/nrf_desktop/src/events/usb_event.h:35]: I also prefer inline, and for the same reason. It's just @divipillai pointed to me the doc style which requires enums to have documentation placed above the values.

Comment (divipillai) [reply applications/nrf_desktop/src/events/usb_event.h:35]: http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#enums - Followed this guideline.

**Comment by greg-fer** [comment]
Forwarded the question about doc building issues to @umapraseeda and @b-gent , who are doc framework contact points.

**Comment by umapraseeda** [comment]
> Forwarded the question about doc building issues to @umapraseeda and @b-gent , who are doc framework contact points.

I think it is because led_id is declared twice in the 2 headers, In led_state_def.h  & led_state.h  . @gmarull, do you suggest adding this warning to [known_warnings.txt](https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/known-warnings.txt)?

**Comment by gmarull** [comment]
> > Forwarded the question about doc building issues to @umapraseeda and @b-gent , who are doc framework contact points.
> 
> I think it is because led_id is declared twice in the 2 headers, in led_state_def.h & led_state.h . @gmarull, do you suggest adding this warning to [known_warnings.txt](https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/known-warnings.txt?rgh-link-date=2022-05-23T08%3A29%3A56Z)?

We cannot have 2 equal declarations in Doxygen. Doxygen has a unique namespace, so one needs to choose which one to include (same as you'd do in an application).

**Comment by pdunaj** [comment]
> > > Forwarded the question about doc building issues to @umapraseeda and @b-gent , who are doc framework contact points.
> > 
> > 
> > I think it is because led_id is declared twice in the 2 headers, in led_state_def.h & led_state.h . @gmarull, do you suggest adding this warning to [known_warnings.txt](https://github.com/nrfconnect/sdk-nrf/blob/main/doc/nrf/known-warnings.txt?rgh-link-date=2022-05-23T08%3A29%3A56Z)?
> 
> We cannot have 2 equal declarations in Doxygen. Doxygen has a unique namespace, so one needs to choose which one to include (same as you'd do in an application).

This is quite problematic. I would like to include enum led_id from this doxygroup but from what you say I see the problem is caused by unrelated apps having enum with the same name declared in doxygen (not even within a group from what I can see).

Hi @gmarull you mention that I should chose which one to include, any hint on how to do it? The group that I included does not even have the enum value that's conflicting. It comes from some other places and indeed is redefined. It just happens that enum name is the same.

**Comment by gmarull** [comment]
In Zephyr we have a special define, `__DOXYGEN__` used to selectively include/exclude things from Doxygen. It's used, for example, to include struct fields that depend on a Kconfig option:

```
#if defined(CONFIG_X) || defined(__DOXYGEN__)
/** Unconditionally show me in docs */
...
#endif
```

I guess you could also use it to hide content from Doxygen.

**Comment by gmarull** [comment]
Note that we should be careful when adding application headers, because of the single namespace issue. Applications are often isolated and don't share code between them, so clashes are more likely. Another option would be to setup Doxygen projects for each application that needs it.

**Comment by pdunaj** [comment]
> In Zephyr we have a special define, `__DOXYGEN__` used to selectively include/exclude things from Doxygen. It's used, for example, to include struct fields that depend on a Kconfig option:
> 
> ```
> #if defined(CONFIG_X) || defined(__DOXYGEN__)
> /** Unconditionally show me in docs */
> ...
> #endif
> ```
> 
> I guess you could also use it to hide content from Doxygen.

I don't think this will work - at least in the long term. We want to document this enum in nrf_desktop and probably would like to do the same in other apps. So excluding that from doxygen would be problematic.

If apps have this enum in specific doxygroups would the problem be fixed? Or would it still pull in all values from different places?

**Comment by pdunaj** [comment]
> Note that we should be careful when adding application headers, because of the single namespace issue. Applications are often isolated and don't share code between them, so clashes are more likely. Another option would be to setup Doxygen projects for each application that needs it.

So far we have not added them but our doc team would like to fill in missing references. I think it was always a question how to construct the documentation for the application but we build one as we get one documentation artifact for nrf, that becomes public. If we generate doc for each app, each app would need to get its own page probably... this will impact whole nrf doc.

**Comment by pdunaj** [comment]
rebased

**Comment by pdunaj** [comment]
Rebased + added a small change that hopefully will get rid of the warnings.
I have tried various things but nothing helps. We simply have to use different name of enums. In our case this is not so problematic but in generic case it will be an issue :(

File: applications/nrf_desktop/configuration/common/led_state.h [hunk 7]
```diff
@@ -4,45 +4,66 @@
...
+	/** Device searches for new peer before bond erase. */
 	LED_PEER_STATE_ERASE_ADV,
 
+	/** Number of peer states. */
 	LED_PEER_STATE_COUNT
 };
 
-enum led_id {
+/** @brief LED identification. */
+enum led_id_nrf_desktop {
```
Comment (pdunaj) [inline applications/nrf_desktop/configuration/common/led_state.h:60]: @divipillai , I added postfix to the name. It's not pretty but should do the trick.

**Comment by pdunaj** [comment]
Hi @divipillai , can you check if it's better now and mergeable?

File: applications/nrf_desktop/README.rst [hunk 8]
```diff
@@ -1857,6 +1857,46 @@ These are valid for events that have many listeners or sources, and are gathered
...
+API documentation
+*****************************
+
+Following are the API elements used by the application configuration.
+
+HID reports
+===========
+
+| Header file: :file:`applications/nrf_desktop/configuration/common/hid_report_desc.h`
+| Source file: :file:`applications/nrf_desktop/configuration/common/hid_report_desc.h`
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1869]:
```suggestion
| Source file: :file:`applications/nrf_desktop/configuration/common/hid_report_desc.c`
```

File: applications/nrf_desktop/README.rst [hunk 9]
```diff
@@ -1857,6 +1857,46 @@ These are valid for events that have many listeners or sources, and are gathered
...
+
+| Header file: :file:`applications/nrf_desktop/configuration/common/led_state.h`
+| Source file: :file:`applications/nrf_desktop/src/modules/led_state.c`
+
+.. doxygengroup:: nrf_desktop_led_state
+   :project: nrf
+   :members:
+
+Application events API
+**********************
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1886]: We could remove this heading as it is aligned to 'API documentation' heading which is bit misleading.

File: applications/nrf_desktop/README.rst [hunk 10]
```diff
@@ -1857,6 +1857,46 @@ These are valid for events that have many listeners or sources, and are gathered
    doc/hfclk_lock.rst
    doc/event_rel_modules.rst
 
+API documentation
```
Comment (divipillai) [inline applications/nrf_desktop/README.rst:1860]: The "API documentation" section could be after the "Application configuration options" section.

File: applications/nrf_desktop/README.rst (hunk 9 above)
Comment (pdunaj) [reply applications/nrf_desktop/README.rst]: done

File: applications/nrf_desktop/README.rst (hunk 10 above)
Comment (pdunaj) [reply applications/nrf_desktop/README.rst:1883]: moved

**Comment by pdunaj** [comment]
Hi @divipillai , first force push changes rst, second moves the docs and final does rebase the PR

**Review by divipillai — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

**Pull request merged**